### PR TITLE
Fix issue of invalid proposals being accepted

### LINKED ISSUE #494: Samplers accept invalid proposals
**Description**

With a bounded prior, invalid proposals are accepted after a MH step. This happens to all samplers involving a MH step.

As seen in the following plot, an invalid proposal, negative here, is accepted after a MH step, which is not supposed to happen, and then the sampler gets stuck there.

**Example to reproduce**

```python
import cuqi
beta = cuqi.distribution.Beta(0.5, 0.5)
simpler = cuqi.sampler.MALA(beta, x0=0.1, scale=0.1)
simples = simpler.sample(1000, 0)
plt.plot(simples.samples.T[:10])
```
bug

# DoD
- [x] Identity the reason behind:
  - [x]  the validity of proposals is not checked at a MH step
- [ ] solution agreed by team and issue merged
  - [x] validate proposal with `isnan` and `isinf` at MH steps in relevant samplers, i.e., MH, CWMH, MALA, NUTS
- [x] proper test functions are created

### PR DESCRIPTION
**Pull request by chaozg**
fixed #494 

With this PR, the following piece of code works now:

```python
import cuqi
import matplotlib.pyplot as plt
import numpy as np

beta = cuqi.distribution.Beta(0.5, 0.5)
sampler = cuqi.experimental.mcmc.MH(beta, initial_point=np.array([0.1]), scale=0.1)
#sampler = cuqi.experimental.mcmc.ULA(beta, initial_point=np.array([0.1]), scale=0.1)
#sampler = cuqi.experimental.mcmc.MALA(beta, initial_point=np.array([0.1]), scale=0.1)
#sampler = cuqi.experimental.mcmc.NUTS(beta, initial_point=np.array([0.1]))
sampler.sample(10)
samples = sampler.get_samples()
samples.plot_trace()
```

MH, CWMH, ULA, MALA and NUTS are also tested for a "vector" Beta distribution with iid components

An irrelevant issue is observed and recorded at https://github.com/CUQI-DTU/CUQIpy/issues/519

### CODE REVIEW

**Review by amal-ghamdi — APPROVED** [review]
Thank you @chaozg! Looks neat and makes sense to me.

**Review by nabriis — CHANGES_REQUESTED** [review]
Hey @chaozg. Thank you for fixing this issue. Could you somehow, based on the `beta` distribution design a test that runs through our list of samplers, sample e.g. 50 samples, then test if any invalid samples are present? This would catch the problem in future implementations. You can try to uncomment our isnan and isinf cases to check if the test would catch the issue.

You can put the test in zexperimental/test_mcmc.py file. You can use one of the existing lists of samplers that we have as a variable in that file. Then try sampling the beta distribution for each of those samplers and compare and check for invalid.

**Comment by chaozg** [comment]
Thank @amal-ghamdi and @nabriis for your review.

@nabriis , I just added a test to check
- if MH, ULA, MALA, NUTS work for scaler Beta(0.5,0.5)
- if MH, MH, CWMH, ULA, MALA, NUTS work for vector iid Beta([0.5, 0.5], [0.5,0.5])

I think I just take a simple (but naive) way to create a test function. Any suggestion to improve this test function is welcome : )

**Comment by chaozg** [comment]
> Hey @chaozg. Thank you for fixing this issue. Could you somehow, based on the `beta` distribution design a test that runs through our list of samplers, sample e.g. 50 samples, then test if any invalid samples are present? This would catch the problem in future implementations. You can try to uncomment our isnan and isinf cases to check if the test would catch the issue.

Yes, test as you suggested and the test case can capture invalid samples, out of bound of [0, 1] in this Beta case.

> 
> You can put the test in zexperimental/test_mcmc.py file. You can use one of the existing lists of samplers that we have as a variable in that file. Then try sampling the beta distribution for each of those samplers and compare and check for invalid.

It seems I can't use the existing samplers as I need to handpick only MH, ULA ,MALA, NUTS for bounded scalar distribution and only MH, CWMH, ULA ,MALA, NUTS for bounded vector distribution. But I might be very wrong here.

File: tests/zexperimental/test_mcmc.py [hunk 1]
```diff
@@ -903,4 +903,33 @@ def test_HybridGibbs_updates_state_only_after_accepting_sample():
...
+        initial_point=np.array([0.1,0.1]))
+]
+
+@pytest.mark.parametrize("sampler", sampler_instances_for_bounded_distribution)
+def test_if_invalid_sample_accepted(sampler: cuqi.experimental.mcmc.Sampler):
+    sampler.sample(50)
+    samples = sampler.get_samples().samples
+    tol = 1e-8
+    assert samples.min() > 0.0 - tol and samples.max() < 1.0 + tol, \
+        f"Invalid samples accepted for sampler {sampler.__class__.__name__}."
```
Comment (nabriis) [inline tests/zexperimental/test_mcmc.py:935]:
```suggestion
sampler_instances_for_bounded_distribution = [
    cuqi.experimental.mcmc.MH(
        target=cuqi.distribution.Beta(0.5, 0.5),
        initial_point=np.array([0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.ULA(
        cuqi.distribution.Beta(0.5, 0.5),
        initial_point=np.array([0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.MALA(
        cuqi.distribution.Beta(0.5, 0.5),
        initial_point=np.array([0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.NUTS(
        cuqi.distribution.Beta(0.5, 0.5), initial_point=np.array([0.1])
    ),
    cuqi.experimental.mcmc.MH(
        target=cuqi.distribution.Beta(np.array([0.5, 0.5]), np.array([0.5, 0.5])),
        initial_point=np.array([0.1, 0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.CWMH(
        target=cuqi.distribution.Beta(
            np.array([0.5, 0.5]), np.array([0.5, 0.5])
        ),
        initial_point=np.array([0.1, 0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.ULA(
        cuqi.distribution.Beta(np.array([0.5, 0.5]), np.array([0.5, 0.5])),
        initial_point=np.array([0.1, 0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.MALA(
        cuqi.distribution.Beta(np.array([0.5, 0.5]), np.array([0.5, 0.5])),
        initial_point=np.array([0.1, 0.1]),
        scale=0.1,
    ),
    cuqi.experimental.mcmc.NUTS(
        cuqi.distribution.Beta(np.array([0.5, 0.5]), np.array([0.5, 0.5])),
        initial_point=np.array([0.1, 0.1]),
    ),
]

@pytest.mark.parametrize("sampler", sampler_instances_for_bounded_distribution)
def test_if_invalid_sample_accepted(sampler: cuqi.experimental.mcmc.Sampler):
    sampler.sample(50)
    samples = sampler.get_samples().samples
    tol = 1e-8
    assert (
        samples.min() > 0.0 - tol and samples.max() < 1.0 + tol
    ), f"Invalid samples accepted for sampler {sampler.__class__.__name__}."
```

Reformatting suggestion.

**Comment by chaozg** [comment]
Thank you @nabriis  for the review and also the hint on using vercel black

**Pull request merged**